test/tests/functional/pbs_cpuset_reliable_job_startup.py

```diff
@@ -0,0 +1,162 @@
+                                id=self.vnA)
+        self.vnA_ncpus = int(rv[0]['resources_available.ncpus'])
+
+        # launch hook
+        self.launch_hook_body = """
+import pbs
+import time
+e=pbs.event()
+pbs.logmsg(pbs.LOG_DEBUG, "Executing launch")
+# print out the vnode_list[] values
+for vn in e.vnode_list:
+    v = e.vnode_list[vn]
+    pbs.logjobmsg(e.job.id, "launch: found vnode_list[" + v.name + "]")
+# print out the vnode_list_fail[] values:
+for vn in e.vnode_list_fail:
+    v = e.vnode_list_fail[vn]
+    pbs.logjobmsg(e.job.id, "launch: found vnode_list_fail[" + v.name + "]")
+if e.job.in_ms_mom():
+    pj = e.job.release_nodes(keep_select="ncpus=%s:mem=%sgb")
+    if pj is None:
+        e.job.Hold_Types = pbs.hold_types("s")
+        e.job.rerun()
+        e.reject("unsuccessful at LAUNCH")
+pbs.logmsg(pbs.LOG_DEBUG, "Sleeping for 20sec")
+time.sleep(20)
+""" % (self.vnA_ncpus, 2)
+
+    def test_reliable_job_startup_on_cpuset(self):
+        """
+        A job is started with two numa nodes and goes in R state.
+        An execjob_launch hook will force job to have only one numa node.
+        The released numa node can be used in another job.
+        """
+
+        # instantiate execjob_launch hook
+        hook_event = "execjob_launch"
+        hook_name = "launch"
+        a = {'event': hook_event, 'enabled': 'true'}
+        stime = int(time.time())
+        self.server.create_import_hook(hook_name, a, self.launch_hook_body)
+
+        # Check mom logs that the launch hook got propagated
+        msg = "Hook;launch.PY;copy hook-related file request received"
+        self.mom.log_match(msg, starttime=stime, interval=16, n=10)
+
+        # Submit job1
+        j = Job(TEST_USER, {
+            ATTR_l + '.select': '2:ncpus=%s:mem=2gb' % self.vnA_ncpus,
+            ATTR_l + '.place': 'free',
+            ATTR_W: 'tolerate_node_failures=job_start'})
+        stime = int(time.time())
+        jid = self.server.submit(j)
+
+        # Check the exec_vnode while in substate 41
+        self.server.expect(JOB, {ATTR_substate: '41'}, id=jid)
+        self.server.expect(JOB, 'exec_vnode', id=jid, op=SET)
+        job_stat = self.server.status(JOB, id=jid)
+        execvnode1 = job_stat[0]['exec_vnode']
+        self.logger.info("initial exec_vnode: %s" % execvnode1)
+        initial_vnodes = execvnode1.split('+')
+
+        # Check the exec_vnode after job is in R state, substate 42
+        self.server.expect(JOB, {ATTR_substate: '42'}, interval=7, id=jid)
+        self.server.expect(JOB, 'exec_vnode', id=jid, op=SET)
+        job_stat = self.server.status(JOB, id=jid)
+        execvnode2 = job_stat[0]['exec_vnode']
+        self.logger.info("pruned exec_vnode: %s" % execvnode2)
+
+        # Check for msg in mom logs
+        msg = "new_cpuset:  setting altid to CPU set named /PBSPro/%s" % jid
+        self.mom.log_match(msg, starttime=stime, regexp=True, n=10)
```

Comment:
I don't think you need regexp=True since the match is for the exact substring.

Reply:
Done.